protocols/v2/roles-logic-sv2/src/channel_logic/channel_factory.rs

```diff
@@ -409,28 +413,28 @@ impl ChannelFactory {
             // If we have just a prev hash we need to send it after the SetupConnectionSuccess
             // message
             (Some((prev_h, _)), None, true) => {
-                let prev_h = prev_h.into_set_p_hash(channel_id);
+                let prev_h = prev_h.into_set_p_hash(channel_id, None);
                 result.push(Mining::SetNewPrevHash(prev_h.clone()));
                 Ok(())
             }
             // If we have a prev hash and a last valid job we need to send before the prev hash and
             // the the valid job
             (Some((prev_h, _)), Some(job), true) => {
-                let prev_h = prev_h.into_set_p_hash(channel_id);
-                result.push(Mining::SetNewPrevHash(prev_h.clone()));
+                let prev_h = prev_h.into_set_p_hash(channel_id, Some(job.job_id));
                 result.push(Mining::NewMiningJob(job));
+                result.push(Mining::SetNewPrevHash(prev_h.clone()));
                 Ok(())
             }
             // If we have everything we need, send before the prev hash and then all the jobs
             (Some((prev_h, _)), Some(job), false) => {
-                let prev_h = prev_h.into_set_p_hash(channel_id);
-                result.push(Mining::SetNewPrevHash(prev_h.clone()));
+                let prev_h = prev_h.into_set_p_hash(channel_id, Some(job.job_id));
```

Comment:
in this case you need to send first the last_valid_job (make sure that you set future=true) then the set_new_prev_hash (make sure that job_id is the one of the previously sented job) the all the other future job that are in self.future_jobs (these job must be sented cause they could be referenced by a future prev_hash)

Reply:
I'm thinking this is already happening. The `last_valid_job` comes from the `future_jobs` list so I'm pretty sure `future_job` would already = true.

My above code will have injected the `last_valid_job`'s id into the SNPH via this `into_set_p_hash(.., Some(job.job_id))` call

Reply:
Here we have:
1. a SNPH
2. a valid job (non future per definition)
3. a list of future job that could be referenced by the next SNPH

Most natural thing to do here would have been to send the SNPH then the non future job then the future jobs. But we added the condition that  after OpenChannel.Success first message received MUST be a future job.

Cuase of that condition what we  need to do is:
1. take the valid job and transform it in a future job and send it as first message
2. send the SNPH (before sending it change the field job_id to match the one of the job sented in (1))
3. send all the future jobs (we still need them cause they are valid future job that can be referenced by a future SNPH)

Reply:
if the future_jobs are sented before the SNPH the downstream will discard all of them but the last_valid_job but this is not what we want

Reply:
done